### PR TITLE
Support for Pumpkaboo/Gourgeist pokemon variants in pokemon json data

### PR DESCRIPTION
**Pull request by magroader**
Focusing on these two because they are the only unsupported pokemon with permanent variants.

### CODE REVIEW

File: converters/pokemon.py [hunk 1]
```diff
@@ -288,20 +298,56 @@ def convert_pdata(input_csv, header=DEFAULT_HEADER):
...
 
             # Each row is one Pokemon
             poke = Pokemon(header)
             poke.setup(row)
+            poke_by_name[poke.name] = poke
+
+        # Some rows are variants of a single pokemon type. Let's go collect those
+        variant_by_type = {}
+        def_poke_by_type = {}
+        def_type_by_variant = {}
```
Comment (Jerakin) [inline converters/pokemon.py]: Would be nice to have the names a bit clearer.
`def` is a keyword in Python which makes the name confusing. A "Bulbasaur" isn't a type of a Pokemon it is a species.

```python
variant_by_species = {}
default_poke_by_species = {}
default_species_by_variant = {}
```

File: converters/pokemon.py [hunk 2]
```diff
@@ -288,20 +298,56 @@ def convert_pdata(input_csv, header=DEFAULT_HEADER):
...
+        # TODO: In the future we may have other variants, like Alolan forms or something.
+        # It's unclear what those might look like from a data perspective
+        for name,variants in util.VARIANT_DATA.items():
```
Comment (Jerakin) [inline converters/pokemon.py]: Missing space after comma

File: converters/pokemon.py [hunk 3]
```diff
@@ -288,20 +298,56 @@ def convert_pdata(input_csv, header=DEFAULT_HEADER):
...
+                        raise Exception("When searching for variants, could not find pokemon of type '" + d["name"] + "'")
+
+        # Merge in all the other variants
+        for type_name, d in variant_by_type.items():
+            def_poke = poke_by_name[def_type_by_variant[def_poke_by_type[type_name]]]
+            variant_poke = poke_by_name[type_name]
+            variant_poke.valid = False
+            def_poke.add_variant(d["variant_name"], variant_poke)
+
+        for name,poke in poke_by_name.items():
```
Comment (Jerakin) [inline converters/pokemon.py]: Missing space after comma

File: converters/pokemon.py [hunk 4]
```diff
@@ -288,20 +298,56 @@ def convert_pdata(input_csv, header=DEFAULT_HEADER):
...
+                        else:
+                            # Store for later, after the default variant has been collected
+                            variant_by_type[d["name"]] = d
+                            def_poke_by_type[d["name"]] = name
+                    else:
+                        raise Exception("When searching for variants, could not find pokemon of type '" + d["name"] + "'")
+
+        # Merge in all the other variants
+        for type_name, d in variant_by_type.items():
+            def_poke = poke_by_name[def_type_by_variant[def_poke_by_type[type_name]]]
```
Comment (Jerakin) [inline converters/pokemon.py]: As the comment about maybe `default_poke` would make it easier to read in the context of python

File: converters/pokemon.py [hunk 5]
```diff
@@ -288,20 +298,56 @@ def convert_pdata(input_csv, header=DEFAULT_HEADER):
...
+                        if "default" in d and d["default"]:
+                            poke.name = name
+                            poke.add_default_variant(d["variant_name"])
+                            def_type_by_variant[name] = d["name"]
+                        else:
+                            # Store for later, after the default variant has been collected
+                            variant_by_type[d["name"]] = d
+                            def_poke_by_type[d["name"]] = name
+                    else:
+                        raise Exception("When searching for variants, could not find pokemon of type '" + d["name"] + "'")
```
Comment (Jerakin) [inline converters/pokemon.py]: Python have quite a few ways format strings, this is my favorite as I find it cleaner and easy to read `raise Exception(f"When searching for variants, could not find pokemon of type {data['name']}")`

File: converters/pokemon.py [hunk 6]
```diff
@@ -288,20 +298,56 @@ def convert_pdata(input_csv, header=DEFAULT_HEADER):
...
+
+        # Some rows are variants of a single pokemon type. Let's go collect those
+        variant_by_type = {}
+        def_poke_by_type = {}
+        def_type_by_variant = {}
+        # TODO: In the future we may have other variants, like Alolan forms or something.
+        # It's unclear what those might look like from a data perspective
+        for name,variants in util.VARIANT_DATA.items():
+            if type(variants) is list:
+                for d in variants:
```
Comment (Jerakin) [inline converters/pokemon.py]: I want bananas with the "naming" comments so might as well make this into a `data` or something more descriptive.

File: util/util.py [hunk 7]
```diff
@@ -91,6 +92,20 @@ def merge(a, b, path=None):
             a[key] = b[key]
     return a
 
+def diff_dict(base, other):
+    diff = {}
+    for k,v in other.items():
```
Comment (Jerakin) [inline util/util.py]: Missing space after comma

**Review by Jerakin — CHANGES_REQUESTED** [review]
Have a few comments, they are mainly naming/style convention stuff. But I would also like to see line 314 to 341 thrown into a function called some thing like `add_variant_info(poke_by_name)` which transforms the Pokemon previously setup, to keep `convert_pdata` as readable as possible.

**Comment by magroader** [comment]
All comments addressed!

**Comment by magroader** [comment]
I just noticed a major issue here - the evolve.json, filter_data.json, and index_order.json are set up incorrectly. I at first though this was just because I didn't download the other .csv files, but it turns out this all came from PDATA. This will need some work!

EDIT: Fixed!

**Comment by Jerakin** [comment]
lgtm

**Review by Jerakin — APPROVED** [review]

**Pull request merged**